File: src/LTAdmin/Services/Infrastructure/BackupService.cs
using LTAdmin.Core;
using LTAdmin.Data;
using LTAdmin.Models.Dto;
using LTAdmin.Services.Common;
using LTAdmin.Services.Logging;
using LTAdmin.Services.Referentiel;

namespace LTAdmin.Services.Infrastructure;

/// <summary>
/// Sauvegardes du fichier LTA_ADM.accdb : copies horodatées dans le dossier
/// « Sauvegardes », inventaire, restauration explicite (avec copie de
/// sécurité préalable) et purge des anciennes copies. La restauration exige
/// une confirmation côté écran : elle remplace le fichier de travail.
/// </summary>
public sealed class BackupService : ServiceBase
{
    public BackupService(AccessDatabase database, AppLogger logger, JournalService journal, ParametreService parametres)
        : base(database, logger, journal, parametres)
    {
    }

    public string BackupsFolder
        => Path.Combine(Path.GetDirectoryName(Db.DatabasePath) ?? AppContext.BaseDirectory, "Sauvegardes");

    /// <summary>Crée une sauvegarde horodatée et retourne son chemin.</summary>
    public Result<string> CreateBackup(string codeUtr)
    {
        try
        {
            if (!File.Exists(Db.DatabasePath))
                return Result<string>.Fail("Base de données introuvable : " + Db.DatabasePath, ErrorCodes.NotFound);
            var destination = Db.CreateBackup();
            var size = new FileInfo(destination).Length;
            if (size == 0)
                return Result<string>.Fail("La sauvegarde créée est vide : vérifiez l’espace disque.", ErrorCodes.Technical);
            Journal.LogOperation(codeUtr, "SAUVEGARDE", null, null, Path.GetFileName(destination));
            Logger.Info("Sauvegarde créée : " + destination);
            return Result<string>.Ok(destination, "Sauvegarde créée : " + Path.GetFileName(destination));
        }
        catch (Exception ex) { return Failure<string>("Sauvegarde de la base", ex); }
    }

    public List<BackupInfo> ListBackups()
    {
        var result = new List<BackupInfo>();
        try
        {
            if (!Directory.Exists(BackupsFolder)) return result;
            foreach (var file in Directory.GetFiles(BackupsFolder, "LTA_ADM_*.accdb"))
            {
                try
                {
                    var info = new FileInfo(file);
                    result.Add(new BackupInfo
                    {
                        FilePath = info.FullName,
                        FileName = info.Name,
                        CreatedAt = info.CreationTime,
                        SizeBytes = info.Length
                    });
                }
                catch { /* Un fichier illisible ne bloque pas l'inventaire. */ }
            }
        }
        catch (Exception ex) { Logger.Error("Inventaire des sauvegardes", ex); }
        return result.OrderByDescending(b => b.CreatedAt).ToList();
    }

    /// <summary>
    /// Restaure une sauvegarde (action explicite irréversible côté données :
    /// une copie de sécurité du fichier courant est conservée avant écrasement).
    /// </summary>
    public Result RestoreBackup(string backupPath, string codeUtr)
    {
        try
        {
            if (string.IsNullOrWhiteSpace(backupPath) || !File.Exists(backupPath))
                return Result.Fail("Fichier de sauvegarde introuvable.", ErrorCodes.NotFound);
            if (!backupPath.EndsWith(".accdb", StringComparison.OrdinalIgnoreCase))
                return Result.Fail("Seuls les fichiers .accdb peuvent être restaurés.", ErrorCodes.Validation);
            if (string.Equals(Path.GetFullPath(backupPath), Path.GetFullPath(Db.DatabasePath), StringComparison.OrdinalIgnoreCase))
                return Result.Fail("Cette sauvegarde est le fichier de travail lui-même.", ErrorCodes.Validation);
            if (Db.InTransaction)
                return Result.Fail("Restauration impossible pendant une transaction.", ErrorCodes.BusinessRule);

            var livePath = Db.DatabasePath;
            Logger.Info($"Restauration demandée par {codeUtr} : {backupPath} -> {livePath}");

            var wasOpen = Db.IsOpen;
            Db.Close();
            try
            {
                // Copie de sécurité du fichier courant avant écrasement.
                if (File.Exists(livePath))
                {
                    Directory.CreateDirectory(BackupsFolder);
                    var safety = Path.Combine(BackupsFolder, $"LTA_ADM_avant_restauration_{DateTime.Now:yyyyMMdd_HHmmss}.accdb");
                    File.Copy(livePath, safety, overwrite: false);
                    Logger.Info("Copie de sécurité avant restauration : " + safety);
                }
                File.Copy(backupPath, livePath, overwrite: true);
            }
            finally
            {
                if (wasOpen) Db.Open();
            }

            Journal.LogOperation(codeUtr, "RESTAURATION", null, null, Path.GetFileName(backupPath));
            return Result.Ok("Base restaurée depuis : " + Path.GetFileName(backupPath));
        }
        catch (Exception ex) { return Failure("Restauration d’une sauvegarde", ex); }
    }

    /// <summary>Supprime les sauvegardes les plus anciennes au-delà du quota conservé.</summary>
    public Result<int> PurgeOldBackups(int keepCount, string codeUtr)
    {
        if (keepCount < 1)
            return Result<int>.Fail("Conservez au moins une sauvegarde.", ErrorCodes.Validation);
        try
        {
            var backups = ListBackups();
            var deleted = 0;
            foreach (var backup in backups.Skip(keepCount))
            {
                try
                {
                    File.Delete(backup.FilePath);
                    deleted++;
                }
                catch (Exception ex)
                {
                    Logger.Warning("Purge impossible pour " + backup.FileName + " : " + ex.Message);
                }
            }
            if (deleted > 0)
                Journal.LogOperation(codeUtr, "PURGE_SAUVEGARDES", null, null, $"{deleted} ancienne(s) sauvegarde(s) supprimée(s).");
            return Result<int>.Ok(deleted, deleted == 0 ? "Aucune sauvegarde à purger." : $"{deleted} sauvegarde(s) purgée(s).");
        }
        catch (Exception ex) { return Failure<int>("Purge des sauvegardes", ex); }
    }
}

File: ltadmin/services/infrastructure/backup_service.py
"""Sauvegardes de la base Access : création, liste, restauration, purge.

Une sauvegarde copie le fichier .accdb fermé (jamais pendant une transaction)
vers le sous-dossier « Sauvegardes ». La restauration conserve d'abord une
copie de sécurité « LTA_ADM_avant_restauration_*.accdb ».
"""

from __future__ import annotations

import datetime as _dt
import os
import shutil
from typing import List, Optional

from ltadmin.core.result import Result
from ltadmin.data.access_database import AccessDatabase
from ltadmin.models.dto import BackupInfo
from ltadmin.services.logging.app_logger import AppLogger

BACKUP_FOLDER_NAME = "Sauvegardes"
SAFETY_PREFIX = "LTA_ADM_avant_restauration_"


class BackupService:

    def __init__(self, database: AccessDatabase, logger: AppLogger):
        self._database = database
        self._logger = logger

    @property
    def backup_directory(self) -> str:
        return os.path.join(
            os.path.dirname(self._database.database_path) or os.getcwd(),
            BACKUP_FOLDER_NAME)

    def create_backup(self) -> Result:
        """Crée une sauvegarde datée et retourne son chemin dans le message."""
        try:
            path = self._database.create_backup()
            return Result.ok(f"Sauvegarde créée : {os.path.basename(path)}")
        except Exception as ex:
            self._logger.error("Création de sauvegarde", ex)
            return Result.fail(
                "Sauvegarde impossible : " + str(ex), "TECHNIQUE")

    def list_backups(self, include_safety_copies: bool = False) -> List[BackupInfo]:
        """Sauvegardes disponibles, de la plus récente à la plus ancienne."""
        result: List[BackupInfo] = []
        try:
            if not os.path.isdir(self.backup_directory):
                return result
            for file_name in os.listdir(self.backup_directory):
                if not file_name.lower().endswith(".accdb"):
                    continue
                if not include_safety_copies and file_name.startswith(SAFETY_PREFIX):
                    continue
                full_path = os.path.join(self.backup_directory, file_name)
                try:
                    stat = os.stat(full_path)
                except OSError:
                    continue
                result.append(BackupInfo(
                    file_path=full_path,
                    file_name=file_name,
                    created_at=_dt.datetime.fromtimestamp(stat.st_mtime),
                    size_bytes=stat.st_size))
            result.sort(key=lambda info: info.created_at or _dt.datetime.min,
                        reverse=True)
        except Exception as ex:
            self._logger.error("Liste des sauvegardes", ex)
        return result

    def restore_backup(self, backup_path: str) -> Result:
        """Restaure une sauvegarde (copie de sécurité préalable de la base)."""
        if not backup_path or not os.path.isfile(backup_path):
            return Result.fail("Fichier de sauvegarde introuvable.", "INTROUVABLE")
        if os.path.abspath(backup_path) == os.path.abspath(self._database.database_path):
            return Result.fail("Choisissez une sauvegarde, pas la base active.", "VALIDATION")
        try:
            os.makedirs(self.backup_directory, exist_ok=True)
            safety_path = os.path.join(
                self.backup_directory,
                f"{SAFETY_PREFIX}{_dt.datetime.now():%Y%m%d_%H%M%S}.accdb")
            was_open = self._database.is_open
            if self._database.in_transaction:
                return Result.fail(
                    "Restauration impossible pendant une transaction.", "GESTION")
            if was_open:
                self._database.close()
            try:
                shutil.copyfile(self._database.database_path, safety_path)
                shutil.copyfile(backup_path, self._database.database_path)
            finally:
                if was_open:
                    self._database.open()
            return Result.ok(
                f"Base restaurée depuis {os.path.basename(backup_path)} "
                f"(copie de sécurité : {os.path.basename(safety_path)}).")
        except Exception as ex:
            self._logger.error("Restauration de sauvegarde", ex)
            return Result.fail("Restauration impossible : " + str(ex), "TECHNIQUE")

    def purge_backups(self, keep_count: int = 10, include_safety_copies: bool = False) -> Result:
        """Supprime les plus anciennes sauvegardes au-delà de keep_count (≥ 1)."""
        keep_count = max(1, keep_count)
        try:
            backups = self.list_backups(include_safety_copies)
            if len(backups) <= keep_count:
                return Result.ok(
                    f"Aucune purge nécessaire ({len(backups)} sauvegarde(s) conservée(s)).")
            removed = 0
            for info in backups[keep_count:]:
                try:
                    os.remove(info.file_path)
                    removed += 1
                except OSError:
                    continue
            return Result.ok(f"{removed} sauvegarde(s) ancienne(s) supprimée(s).")
        except Exception as ex:
            self._logger.error("Purge des sauvegardes", ex)
            return Result.fail("Purge impossible : " + str(ex), "TECHNIQUE")

    @staticmethod
    def format_size(size_bytes: int) -> str:
        if size_bytes >= 1024 * 1024:
            return f"{size_bytes / (1024 * 1024):.1f} Mo"
        if size_bytes >= 1024:
            return f"{size_bytes / 1024:.1f} Ko"
        return f"{size_bytes} o"

    @staticmethod
    def describe(info: BackupInfo) -> str:
        created = info.created_at.strftime("%d/%m/%Y %H:%M:%S") if info.created_at else "?"
        return f"{info.file_name} — {created} — {BackupService.format_size(info.size_bytes)}"
